{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a41cb555-542a-4c7a-b0b9-507adc019855",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.legend import Legend\n",
    "import matplotlib.colors as colors\n",
    "from matplotlib import cm\n",
    "from mpl_toolkits.mplot3d import axes3d\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import scanpy as sc\n",
    "\n",
    "from sklearn import datasets\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "from numba import jit\n",
    "\n",
    "import celltypist\n",
    "from celltypist import models\n",
    "\n",
    "from matplotlib.cm import ScalarMappable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ff0eace-80cf-4956-84aa-462161d87bb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Custom colormap\n",
    "\n",
    "from matplotlib.cm import register_cmap\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "tab20b = matplotlib.colormaps['tab20b']\n",
    "tab20c = matplotlib.colormaps['tab20c']\n",
    "colors1 = tab20b(np.linspace(3.001/5., 1, 8))\n",
    "colors2 = tab20c(np.linspace(0, 3.999/5., 16))\n",
    "\n",
    "colors = np.concatenate([colors1, colors2])\n",
    "\n",
    "map_name = 'op_tab24'\n",
    "op_cmap = ListedColormap(colors, name=map_name )\n",
    "matplotlib.colormaps.register(name=map_name, cmap=op_cmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87a1fc6e-05da-4adb-b841-ddca15ac1761",
   "metadata": {},
   "outputs": [],
   "source": [
    "from mpl_toolkits.axes_grid1 import AxesGrid\n",
    "\n",
    "def shiftedColorMap(cmap, start=0, midpoint=0.5, stop=1.0, name='shiftedcmap'):\n",
    "    '''\n",
    "    Function to offset the \"center\" of a colormap. Useful for\n",
    "    data with a negative min and positive max and you want the\n",
    "    middle of the colormap's dynamic range to be at zero.\n",
    "\n",
    "    Input\n",
    "    -----\n",
    "      cmap : The matplotlib colormap to be altered\n",
    "      start : Offset from lowest point in the colormap's range.\n",
    "          Defaults to 0.0 (no lower offset). Should be between\n",
    "          0.0 and `midpoint`.\n",
    "      midpoint : The new center of the colormap. Defaults to \n",
    "          0.5 (no shift). Should be between 0.0 and 1.0. In\n",
    "          general, this should be  1 - vmax / (vmax + abs(vmin))\n",
    "          For example if your data range from -15.0 to +5.0 and\n",
    "          you want the center of the colormap at 0.0, `midpoint`\n",
    "          should be set to  1 - 5/(5 + 15)) or 0.75\n",
    "      stop : Offset from highest point in the colormap's range.\n",
    "          Defaults to 1.0 (no upper offset). Should be between\n",
    "          `midpoint` and 1.0.\n",
    "    '''\n",
    "    cdict = {\n",
    "        'red': [],\n",
    "        'green': [],\n",
    "        'blue': [],\n",
    "        'alpha': []\n",
    "    }\n",
    "\n",
    "    # regular index to compute the colors\n",
    "    reg_index = np.linspace(start, stop, 257)\n",
    "\n",
    "    # shifted index to match the data\n",
    "    shift_index = np.hstack([\n",
    "        np.linspace(0.0, midpoint, 128, endpoint=False), \n",
    "        np.linspace(midpoint, 1.0, 129, endpoint=True)\n",
    "    ])\n",
    "\n",
    "    for ri, si in zip(reg_index, shift_index):\n",
    "        r, g, b, a = cmap(ri)\n",
    "\n",
    "        cdict['red'].append((si, r, r))\n",
    "        cdict['green'].append((si, g, g))\n",
    "        cdict['blue'].append((si, b, b))\n",
    "        cdict['alpha'].append((si, a, a))\n",
    "\n",
    "    newcmap = matplotlib.colors.LinearSegmentedColormap(name, cdict)\n",
    "    plt.register_cmap(cmap=newcmap)\n",
    "\n",
    "    return newcmap\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "588eb4d7-f3d9-451a-a6e6-63fa9c1149f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load preprocessed data\n",
    "gspDF = pd.read_csv('/Users/oipulk/Documents/prime_data/preprocessed/gsp_clin_noiRECISTNans.csv',low_memory=False, index_col=0)\n",
    "pspDF = pd.read_csv('/Users/oipulk/Documents/prime_data/preprocessed/psp_clin_noiRECISTNans.csv',low_memory=False, index_col=0)\n",
    "\n",
    "gsp_genes = gspDF.index.values[13:len(gspDF.index.values)]\n",
    "psp_genes = pspDF.index.values[13:len(pspDF.index.values)]\n",
    "ctl_genes = ['CD8A', 'CD8B', 'GZMA', 'GZMB', 'PRF1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82cc63e0-55eb-4186-8cc6-34cf3d4bbbeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Genes tested in laboratory and their p-value in t-test for response\n",
    "\n",
    "orderedGenes = np.array(['SLAMF7', 'ATP2A3', 'GBP1', 'HSH2D', 'LCK', 'MAP4K1', 'CSF2RB', 'CD247', 'NCF1', 'LCP2', 'PRKCH', 'ITGAL', 'LY9', 'HLA-DPA1', 'TIGIT', 'CCDC69', 'RASAL3', 'IL12RB1', 'SASH3', 'IL2RG', 'ARHGAP9', 'ARHGAP25', 'PTK2B', 'TYMP', 'PSD4', 'WAS', 'CD27', 'SLAMF6', 'TMC8', 'GBP5', 'SEPT1', 'TRAF3IP3', 'PSMB10', 'GZMB', 'CD3E', 'SYK', 'PSTPIP1', 'GNLY', 'MYO1G', 'CD40', 'POU2F2', 'CD38', 'SEMA4D', 'FCRL5', 'IGFLR1', 'SIRPG', 'BTK', 'ITGB2', 'HCLS1', 'CD48', 'AOAH', 'RCSD1', 'GPR174', 'SEMA4A', 'PARVG', 'PRF1', 'DOCK2', 'FGR', 'IL10RA', 'FAM107B', 'EVL', 'BIN2', 'HLA-DOB', 'CD52', 'LST1', 'CD84', 'CD53', 'PIK3AP1', 'CLEC2D', 'LYN', 'PTPRC', 'CD3D', 'CD96', 'DOK2', 'TRIM22', 'FGD2', 'EPSTI1', 'CD19', 'VNN2', 'PIK3CG', 'CASP10', 'CD79A', 'GZMA', 'CD2', 'HLA-F', 'MZB1', 'ERAP1', 'CD37', 'FERMT3', 'PLCG2', 'CCR5', 'BTN3A3', 'ZAP70', 'SELL', 'HAVCR2', 'FLT3LG', 'CTSS', 'TRAT1', 'TNFAIP2', 'BLNK', 'ARHGAP30', 'TTC7A', 'SPN', 'CTLA4', 'GIMAP7', 'GBP3', 'LCP1', 'TAPBP', 'CLIC2', 'CYTH4', 'SLC7A7', 'TNFRSF1B', 'BTN3A1', 'LILRB2', 'SEPT6', 'LAT2', 'CCR2', 'CD4', 'NFAM1', 'TAPBPL', 'IGSF6', 'SECTM1', 'UBA7', 'IDO1', 'RSAD2', 'CYBB', 'PTAFR', 'PLCB2', 'ADAM28', 'SELPLG', 'TMEM176B', 'PTPN22', 'NAAA', 'HCK', 'HLA-DQB1', 'RASGEF1B', 'TBXAS1', 'LSP1', 'ADAM8', 'MPEG1', 'TNFSF10', 'PLD4', 'LILRB4', 'STK17B', 'ARRB2', 'SLCO2B1', 'TNFSF13', 'SLAMF1', 'CD28', 'SLC15A3', 'ITGAM', 'RASGRP2', 'VCAM1', 'CLEC7A', 'HLA-DMB', 'CTSW', 'HCST', 'SLC8A1', 'SKAP1', 'IL2RB', 'BASP1', 'RNASE6', 'LGMN'\n",
    ",'PIK3IP1', 'CYTH1', 'ALDH2', 'GLIPR1', 'ELMO1', 'EVI2B', 'MFNG', 'VAMP5', 'LILRA4', 'MAN1A1', 'KLHL6', 'CD1C', 'CD163', 'MS4A6A', 'ICAM2', 'ACE', 'FGL2', 'NCKAP1L', 'KCNMA1', 'SRGN', 'CST7', 'DAPP1', 'FOLR2', 'ENTPD1', 'PILRA', 'HLA-DMA', 'SAMSN1', 'LAMP3', 'CD14', 'PDCD1LG2', 'CLEC10A', 'GNA15', 'ADAP2', 'SPINT2', 'CD33', 'CR1', 'CXCL12', 'IL7R', 'AIF1', 'PTGER4', 'EMB', 'FAM57A', 'SLC39A10'])\n",
    "\n",
    "\n",
    "orderedGenes_p = np.array([0.000002, 0.000171, 0.00026, 0.000262, 0.000288, 0.000316, 0.000355, 0.000377, 0.000445, 0.00051, 0.000515, 0.000672, 0.000698, 0.000757, 0.000763, 0.000841, 0.000858, 0.000889, 0.000987, 0.00131, 0.00138, 0.001427, 0.001449, 0.001493, 0.001516, 0.001584, 0.001614, 0.001712, 0.001857, 0.00186, 0.001904, 0.002429, 0.002451, 0.002531, 0.002572, 0.002623, 0.002647, 0.002652, 0.002658, 0.002798, 0.003069, 0.00316, 0.003734, 0.003758, 0.004254, 0.00434, 0.004603, 0.004735, 0.004918, 0.005094, 0.005575, 0.005847, 0.006044, 0.006048, 0.006177, 0.006216, 0.006241, 0.006269, 0.00664, 0.006877, 0.007022, 0.007058, 0.007183, 0.007487, 0.008256, 0.008742, 0.008766, 0.008771, 0.009672, 0.009696, 0.009935, 0.01011, 0.010196, 0.010751, 0.011399, 0.011402, 0.011679, 0.011761, 0.012181, 0.012613, 0.012746, 0.012924, 0.012959, 0.013217, 0.013969, 0.014275, 0.015414, 0.015982, 0.017934, 0.018331, 0.018413, 0.018533, 0.019213, 0.020528, 0.020969, 0.021716, 0.022919, 0.024311, 0.02618, 0.026245, 0.026588, 0.027259, 0.027294, 0.027622, 0.027867, 0.027916, 0.029668, 0.030289, 0.030526, 0.031287, 0.031835, 0.032011\n",
    ", 0.032486, 0.033289, 0.033758, 0.034644, 0.034706, 0.03764, 0.038836, 0.0441, 0.044157, 0.049343, 0.053912, 0.058147, 0.058859, 0.05891, 0.063574, 0.064643, 0.064875, 0.070647, 0.07081, 0.071064, 0.071227, 0.077519, 0.080847, 0.083266, 0.085183, 0.091863, 0.100711, 0.102939, 0.105291, 0.110201, 0.119018, 0.125466, 0.127341, 0.140282, 0.143816, 0.146222, 0.149913, 0.15029, 0.151847, 0.162302, 0.162963, 0.162976, 0.166283, 0.1701, 0.17621, 0.179118, 0.199627, 0.202331, 0.220066, 0.243455, 0.244815, 0.24494, 0.250183, 0.268815, 0.270333, 0.270414, 0.283139, 0.285096, 0.300669, 0.304284, 0.319236, 0.329626, 0.347148, 0.356216, 0.362608, 0.365105, 0.368516, 0.387825, 0.388237, 0.412144, 0.414301, 0.41613, 0.468657, 0.491555, 0.51177, 0.54486, 0.564074, 0.5938, 0.60468, 0.615798, 0.636671, 0.65277, 0.657607, 0.661587, 0.701709, 0.721311, 0.756649, 0.778583, 0.819239, 0.837958, 0.86186, 0.955137, 0.001639, 0.068734])\n",
    "\n",
    "orderedGenes = orderedGenes[np.argsort(orderedGenes_p)]\n",
    "orderedGenes_p = np.sort(orderedGenes_p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "445d249a-8071-472f-bab4-130072993bee",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata= sc.read_loom(\n",
    "    \"/Users/oipulk/Documents/scRNASeq/data/Li2018/Li.loom\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1640f13b-8042-4b50-bbf8-0d2797598a1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell type predictions by Celltypist\n",
    "predictions = celltypist.annotate(adata, model = 'Immune_All_Low.pkl', majority_voting = True)\n",
    "print(predictions.predicted_labels)\n",
    "adata = predictions.to_adata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7158fa05-aaf6-4b21-bde6-14a2b1284596",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['target.y'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b790af0-25e9-471a-b345-c603ac08cba0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute frequencies of anti-MAA target counts in each patient (normalized by total number of target annotations, and all cell counts)\n",
    "patients = np.unique(adata.obs['patient'].values)\n",
    "anti_maa_f = np.zeros(adata.shape[0])\n",
    "anti_maa_f_all = np.zeros(adata.shape[0])\n",
    "\n",
    "for pat in patients:\n",
    "\n",
    "    # Targets are anti-melanoma, anti-viral, Multi, NA, or None. Compute proportion of anti-melanoma of designated targets.\n",
    "    if (sum(adata.obs['patient']==pat)!=(sum(adata[adata.obs['patient']==pat].obs['target.y']=='NA')+sum(adata[adata.obs['patient']==pat].obs['target.y']=='None'))):\n",
    "        \n",
    "        prop = sum(adata[adata.obs['patient']==pat].obs['target.y']=='anti-melanoma')/(sum(adata.obs['patient']==pat)-sum(adata[adata.obs['patient']==pat].obs['target.y']=='NA')-sum(adata[adata.obs['patient']==pat].obs['target.y']=='None'))\n",
    "    \n",
    "    else:\n",
    "\n",
    "        prop=np.nan\n",
    "    \n",
    "    prop_all = sum(adata[adata.obs['patient']==pat].obs['target.y']=='anti-melanoma')/sum(adata.obs['patient']==pat)\n",
    "    anti_maa_f[np.where(adata.obs['patient']==pat)[0]] = prop\n",
    "    anti_maa_f_all[np.where(adata.obs['patient']==pat)[0]] = prop_all\n",
    "\n",
    "adata.obs['anti_maa_f'] = anti_maa_f\n",
    "adata.obs['anti_maa_f_all'] = anti_maa_f_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf90b939-3a23-4e69-abd6-1ffd40380d54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute CTL score in each patient\n",
    "ctl_genes = ['CD8A', 'CD8B', 'GZMA', 'GZMB', 'PRF1']\n",
    "sc.tl.score_genes(adata, ctl_genes, score_name='ctl_score')\n",
    "\n",
    "pt_ctl_scores = np.zeros(adata.shape[0])\n",
    "for pat in patients:\n",
    "\n",
    "    score = np.nanmean(adata[adata.obs['patient']==pat].obs['ctl_score'])\n",
    "    pt_ctl_scores[np.where(adata.obs['patient']==pat)[0]] = score\n",
    "\n",
    "adata.obs['ctl_score_pt'] = pt_ctl_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea85bb10-50c7-483a-aa36-d4b3c809c79e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add an indicator for prior immunotherapy\n",
    "IT = (adata.obs['treatment'].values=='1IT')|(adata.obs['treatment'].values=='2IT')|(adata.obs['treatment'].values=='mIT').astype(int)\n",
    "adata.obs['IT'] = IT \n",
    "adata.obs['IT'] = adata.obs['IT'].astype('category')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78a43431-a7f3-446f-aff2-d3bd4f19818b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Therapy-naive patients\n",
    "adata_th_naive = adata[adata.obs['IT']==0].copy()\n",
    "\n",
    "# Normalize counts log-transform again\n",
    "sc.pp.normalize_total(adata_th_naive, target_sum=1e4)\n",
    "sc.pp.log1p(adata_th_naive)\n",
    "print(np.expm1(adata_th_naive.X).sum(axis=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b383ee0-475d-44de-b8ca-4a66bd007e0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_th_naive.obs['anti_maa_f'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41ed6167-9425-45d4-a18f-c1ac158f77fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute umap and t-sne\n",
    "sc.pp.pca(adata_th_naive)\n",
    "sc.pp.neighbors(adata_th_naive)\n",
    "sc.tl.umap(adata_th_naive)\n",
    "sc.tl.tsne(adata_th_naive)\n",
    "umap_coordinates = adata_th_naive.obsm['X_umap']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c3fdd67-af72-4352-88b8-de5ed66e4398",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(scanpy=True, dpi=300, dpi_save=1200, frameon=True, vector_friendly=True, fontsize=8,\n",
    "                         figsize=(9,8),  format='pdf', facecolor=None, transparent=False, ipython_format='png2x')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44d70716-1b78-4bea-98d7-e34e77d77278",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First check the current categories\n",
    "print(\"Current categories:\", adata_th_naive.obs['majority_voting'].cat.categories)\n",
    "\n",
    "# ALveolar macrophages is wrong -> Suppressor macrophages (annotation by ACT based on DEGs)\n",
    "adata_th_naive.obs['majority_voting'] = adata_th_naive.obs['majority_voting'].cat.rename_categories({\n",
    "    'Alveolar macrophages': 'Suppressor macrophages'\n",
    "})\n",
    "\n",
    "# Verify the changes\n",
    "print(\"Updated categories:\", adata_th_naive.obs['majority_voting'].cat.categories)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a7f18cf-940e-4f83-9207-ecbbb20a8f50",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.tsne(adata_th_naive, color = 'majority_voting', palette='op_tab24', legend_loc = 'on data',save='Annotations_all_types.pdf' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b526b4ee-9a7a-4854-9cbd-cfb4e7661a9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.tsne(adata_th_naive, color = 'majority_voting', palette='op_tab24', legend_loc = 'right margin',save='Annotations_all_types_legend_right.pdf' )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbd417a8-b444-4f1b-95cc-1dd3094fc692",
   "metadata": {},
   "source": [
    "## Analysis of monocytes, macrophages, DCs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81ee1f7a-d78a-48e5-b678-5beb7590abfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find communities of cells (cell types)\n",
    "sc.tl.leiden(adata_th_naive, key_added='leiden_initial')\n",
    "# sc.tl.louvain(adata_th_naive)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "884d2df8-7d3c-409d-8e64-246f4b78bcc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot UMAP with colors for each cell type by 'leiden' community finding algorithm\n",
    "sc.set_figure_params(scanpy=True, dpi=300, dpi_save=1200, frameon=True, vector_friendly=True, fontsize=8,\n",
    "                         figsize=(9,8),  format='pdf', facecolor=None, transparent=False, ipython_format='png2x')\n",
    "\n",
    "sc.pl.tsne(adata_th_naive, color='leiden_initial', palette='op_tab24', legend_loc='on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50871767-33eb-4fbc-a623-d741b8150c67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Iitial clustering\n",
    "\n",
    "# 2. Identify the cluster(s) containing DC2, monocytes, and macrophages\n",
    "target_clusters = ['6','8','13']  # Replace with your actual cluster IDs\n",
    "\n",
    "# 3. Subset the data\n",
    "adata_subset = adata_th_naive[adata_th_naive.obs['leiden_initial'].isin(target_clusters)].copy()\n",
    "\n",
    "# 4. Recompute the neighborhood graph on the subset\n",
    "sc.pp.neighbors(adata_subset)\n",
    "\n",
    "# 5. Perform Leiden clustering at higher resolution on the subset\n",
    "sc.tl.leiden(adata_subset, resolution=0.75, key_added='leiden_refined')\n",
    "\n",
    "# 6. Prepare categories for the combined clustering\n",
    "initial_categories = list(adata_th_naive.obs['leiden_initial'].cat.categories)\n",
    "refined_categories = list(adata_subset.obs['leiden_refined'].cat.categories)\n",
    "\n",
    "# Remove target clusters from initial categories\n",
    "initial_categories_filtered = [cat for cat in initial_categories if cat not in target_clusters]\n",
    "\n",
    "# Create new category names for refined clusters\n",
    "refined_categories_renamed = [f'r{cat}' for cat in refined_categories]\n",
    "\n",
    "# Combine filtered initial categories with renamed refined categories\n",
    "combined_categories = initial_categories_filtered + refined_categories_renamed\n",
    "\n",
    "# 7. Create new column for combined clustering\n",
    "adata_th_naive.obs['leiden'] = pd.Categorical(\n",
    "    adata_th_naive.obs['leiden_initial'],\n",
    "    categories=combined_categories\n",
    ")\n",
    "\n",
    "# 8. Update the combined clustering for the refined subset\n",
    "for idx in adata_subset.obs.index:\n",
    "    refined_value = adata_subset.obs.loc[idx, 'leiden_refined']\n",
    "    adata_th_naive.obs.at[idx, 'leiden'] = f'r{refined_value}'\n",
    "\n",
    "# 9. Optionally, sort the categories for better readability\n",
    "adata_th_naive.obs['leiden'] = adata_th_naive.obs['leiden'].cat.reorder_categories(sorted(adata_th_naive.obs['leiden'].cat.categories))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "449a54ad-af4f-40a1-b686-2b5972dca56f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.tsne(adata_th_naive, color='leiden', palette='op_tab24', legend_loc='on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78cdc548-c3fc-46c5-a209-6208c55d60d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check that all clusters express either CD45 or CD3 (i.e. they are immune cells)\n",
    "sc.pl.tsne(adata_th_naive, color=['CD45','CD3'],palette='tab20',cmap='coolwarm',vmax=2.0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0052e9a0-7a06-48b2-825d-0a74db5fe977",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_subset = adata_th_naive[adata_th_naive.obs['leiden'].isin(['r0','r1','r2','r3','r4','r5','r6','r7','r8','16','17'])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47e26b28-816b-47f7-b9b4-e8d1d211d57b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell type predictions by Celltypist\n",
    "predictions_subset = celltypist.annotate(adata_subset, model = 'Immune_All_High.pkl', majority_voting = True)\n",
    "print(predictions_subset.predicted_labels)\n",
    "adata_subset = predictions_subset.to_adata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92fd9de0-a0ae-4f32-b5cf-984a083f742d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.tsne(adata_subset, color = 'majority_voting', palette='tab20',legend_loc = 'right margin' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ee16a78-e2d7-4278-9095-1c9546d8bf86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Markers from ACT human pan-tissue (frequency>5)\n",
    "pDC_markers = ['IL3RA','LILRA4','PLD4']\n",
    "sc.tl.score_genes(adata_subset, pDC_markers, score_name='pDC_score')\n",
    "sc.pl.tsne(adata_subset, color=['pDC_score'],palette='tab20',cmap='coolwarm',vmax=2)\n",
    "\n",
    "# cDC1_markers = ['CLEC9A','XCR1','BATF3','CADM1','CD1C'\n",
    "# ,'IDO1','ANXA6','CD40','CLIC2','CST7','ETV3','ID2','IL6ST','NET1','PTDSS1'\n",
    "# ,'SCARB1','TNFRSF10B']\n",
    "\n",
    "cDC1_markers = ['CLEC9A','XCR1','BATF3','IRF8']\n",
    "\n",
    "sc.tl.score_genes(adata_subset, cDC1_markers, score_name='cDC1_score')\n",
    "sc.pl.tsne(adata_subset, color=['cDC1_score'],palette='tab20',cmap='coolwarm',vmax=1.0)\n",
    "\n",
    "# cDC2_markers = ['CD1C','CLEC10A','FCER1A','SIRPA','ADAM8','ADAM28','ARAF','AXL','FCGR2B','HMGA1','IRF4','LY86','PFDN1','TIMM13']\n",
    "\n",
    "cDC2_markers = ['CD1C','CLEC10A','IRF4', 'NOTCH2']\n",
    "\n",
    "sc.tl.score_genes(adata_subset, cDC2_markers, score_name='cDC2_score')\n",
    "sc.pl.tsne(adata_subset, color=['cDC2_score'],palette='tab20',cmap='coolwarm',vmax=1.5)\n",
    "\n",
    "#Langerhans cells\n",
    "LC_markers = ['CD207', 'CD1A', 'CDH1', 'EPCAM']\n",
    "sc.tl.score_genes(adata_subset, LC_markers, score_name='LC_score')\n",
    "sc.pl.tsne(adata_subset, color=['LC_score'],palette='tab20',cmap='coolwarm',vmax=1.5)\n",
    "\n",
    "migratory_state_markers = ['CCR7', 'LAMP3','HLA-DRA']\n",
    "sc.tl.score_genes(adata_subset, migratory_state_markers, score_name='migratory_state_score')\n",
    "sc.pl.tsne(adata_subset, color=['migratory_state_score'],palette='tab20',cmap='coolwarm',vmax=3.5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4422b673-e241-4dc3-9db9-e265f7ae17ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#individual markers\n",
    "sc.pl.tsne(adata_subset, color=['CD14','SIRPA','XCR1'],palette='tab20',cmap='coolwarm')\n",
    "sc.pl.tsne(adata_subset, color=['CD40', 'CD80', 'CD86'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93969343-911b-4d07-8dda-c0d470b315dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata_th_naive,'leiden', method='wilcoxon')\n",
    "sc.pl.rank_genes_groups(adata_th_naive, n_genes=20, sharey=False)\n",
    "\n",
    "de_markers = sc.get.rank_genes_groups_df(adata_th_naive, None)\n",
    "de_markers = de_markers[(de_markers.pvals_adj < 0.03) & (de_markers.logfoldchanges > 1.0)]\n",
    "de_markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69ebe0a2-cce0-4b96-8fe4-6a58db66dd1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For exporting to ACT\n",
    "for cluster in ['r0','r1','r2','r3','r4','r5','r6','r7','r8','16','17']:\n",
    "    print(cluster,':'+', '.join(de_markers[de_markers['group']==cluster].iloc[np.argsort(de_markers[de_markers['group']==cluster]['scores'])[::-1],:][0:50].names.values))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69889f16-e57c-4ccf-8ceb-325c513bb61e",
   "metadata": {},
   "source": [
    "### Compute PRIME scores of cells and patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ba7fcb3-dec6-45c4-a64a-05703d0c01fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Prime score \n",
    "PRIME_df= pd.read_csv('/Users/oipulk/Documents/prime_v3/MS/pub_tables/bDEA_results_and_PRIME_weights.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1548e676-dfd9-483a-b38f-661b112c0f5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "PRIME_df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "942dc96f-ee7e-4a72-8d38-3015db483ed9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#LAP3 (or its aliases) is not in the data\n",
    "PRIME_weights = np.append(PRIME_df.weights.values[0:9],PRIME_df.weights.values[10:15])\n",
    "PRIME_genes = np.append(PRIME_df.index[0:9],PRIME_df.index[10:15])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a43c91f-776e-4c7e-a413-e7b05229dcea",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Check the correct format: PRIME uses log2-transformed TPM data\n",
    "print(np.expm1(adata_th_naive.X).sum(axis=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22775387-f010-40d5-bf0b-8ebe7b72997f",
   "metadata": {},
   "outputs": [],
   "source": [
    "xpr_data = adata_th_naive.copy()\n",
    "\n",
    "## Here we do not need to take the log again\n",
    "# PRIME_score = np.asarray(np.matmul(PRIME_weights,np.log2(1+xpr_data[:,PRIME_genes].X.todense()).T))[0]\n",
    "PRIME_score = np.asarray(np.matmul(PRIME_weights,xpr_data[:,PRIME_genes].X.todense().T))[0]\n",
    "\n",
    "adata_th_naive.obs['PRIME_score'] = PRIME_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2355923-e60c-401d-ad25-7f4f0b6c156f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt_prime_scores = np.zeros(adata_th_naive.shape[0])\n",
    "for pat in patients:\n",
    "\n",
    "    score = np.nanmean(adata_th_naive[adata_th_naive.obs['patient']==pat].obs['PRIME_score'])\n",
    "    pt_prime_scores[np.where(adata_th_naive.obs['patient']==pat)[0]] = score\n",
    "\n",
    "adata_th_naive.obs['PRIME_score_pt'] = pt_prime_scores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bf0f5af-8350-45ca-877c-97182126f4d3",
   "metadata": {},
   "source": [
    "## Anti-MAA split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7492877e-0523-4a84-9c03-9405c54efcb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "anti_maa_pts = np.unique(adata_th_naive.obs[~np.isnan(adata_th_naive.obs['anti_maa_f'])]['patient'].values)\n",
    "adata_amaa = adata_th_naive[adata_th_naive.obs['patient'].isin(anti_maa_pts)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "937d82c9-91a8-4ba0-bb9e-b16348206f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_amaa.obs['anti_maa_hi'] = adata_amaa.obs['anti_maa_f']>np.quantile(np.unique(adata_amaa.obs['anti_maa_f']), [0.25,0.5,0.75])[1]\n",
    "adata_amaa.obs['anti_maa_q1'] = adata_amaa.obs['anti_maa_f']>np.quantile(np.unique(adata_amaa.obs['anti_maa_f']), [0.25,0.5,0.75])[2]\n",
    "adata_amaa.obs['anti_maa_q4'] = adata_amaa.obs['anti_maa_f']<=np.quantile(np.unique(adata_amaa.obs['anti_maa_f']), [0.25,0.5,0.75])[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3202be23-0b9f-401b-a82f-02f7c7485d86",
   "metadata": {},
   "source": [
    "###  PRIME score, SLAMF7, TYMP, and CD74 in high/low anti-MAAcohorts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01f599d8-c048-4d6f-962d-b18f3ad95ebb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from matplotlib.colors import to_rgb\n",
    "from matplotlib.collections import PolyCollection\n",
    "from matplotlib.legend_handler import HandlerTuple\n",
    "import textwrap\n",
    "\n",
    "def compare_gene_expression(df, ordered_cell_list, features, condition_key, condition1, condition2, cell_type_key):\n",
    "    \"\"\"\n",
    "    Compare gene expression between conditions across multiple cell types and genes,\n",
    "    including an analysis of all cells combined.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas DataFrame\n",
    "        Input data frame containing expression data\n",
    "    ordered_cell_list : list\n",
    "        List of cell types in desired order\n",
    "    features : list\n",
    "        List of genes or signature scores to analyze\n",
    "    condition_key : str\n",
    "        Column name for condition \n",
    "    condition1: str\n",
    "        Name of condition 1 (e.g. 'wt')\n",
    "    condition2: str\n",
    "        Name of condition 2 (e.g. 'ko')       \n",
    "    cell_type_key : str\n",
    "        Column name for cell type (e.g. 'leiden' or 'cell_type')\n",
    "    \n",
    "    Returns:\n",
    "    --------\n",
    "    tuple: (pandas DataFrame with detailed results, pandas DataFrame with pivoted results)\n",
    "    \"\"\"\n",
    "    \n",
    "    cell_types = pd.Categorical(list(ordered_cell_list))\n",
    "    \n",
    "    all_results = []\n",
    "    all_p_values = []  # Store all p-values for global correction\n",
    "    \n",
    "    # First, calculate all p-values\n",
    "    for cell_type in cell_types:\n",
    "        \n",
    "        # For 'Combined Cells', use the complete dataset; otherwise filter by cell type\n",
    "        if cell_type == 'Combined Cells':\n",
    "            cell_type_data = df\n",
    "        \n",
    "        else:\n",
    "            cell_type_data = df[df[cell_type_key] == cell_type]\n",
    "\n",
    "        direction = '_'\n",
    "        \n",
    "        for feature in features:\n",
    "            \n",
    "            c1_expr = cell_type_data[cell_type_data[condition_key] ==  condition1][feature]\n",
    "            c2_expr = cell_type_data[cell_type_data[condition_key] ==  condition2][feature]\n",
    "            \n",
    "            # Calculate p-value if enough samples\n",
    "            if (len(c1_expr) > 1) & (len(c2_expr) > 1):\n",
    "                statistic, p_value = stats.mannwhitneyu(c1_expr, c2_expr, alternative='two-sided')\n",
    "            else:\n",
    "                p_value = 1.0\n",
    "\n",
    "            if (len(c1_expr) > 0) & (len(c2_expr) > 0):\n",
    "                \n",
    "                if np.mean(c1_expr) > np.mean(c2_expr):\n",
    "                    direction = '>'\n",
    "    \n",
    "                if np.mean(c1_expr) < np.mean(c2_expr):\n",
    "                    direction = '<'\n",
    "                \n",
    "            all_results.append({\n",
    "                'cell_type': cell_type,\n",
    "                'feature': feature,\n",
    "                'direction': direction,\n",
    "                'p_value': p_value\n",
    "            })\n",
    "            all_p_values.append(p_value)\n",
    "    \n",
    "    # Convert to DataFrame\n",
    "    results_df = pd.DataFrame(all_results)\n",
    "    \n",
    "    # Correct all p-values together\n",
    "    _, p_values_corrected, _, _ = multipletests(all_p_values, method='fdr_bh')\n",
    "    results_df['p_value_corrected'] = p_values_corrected\n",
    "    \n",
    "    # Add significance asterisks based on corrected p-values\n",
    "    results_df['significance'] = results_df['p_value_corrected'].apply(\n",
    "        lambda p: '****' if p < 0.0001 else\n",
    "        ('***' if p < 0.001 else\n",
    "        ('**' if p < 0.01 else\n",
    "        ('*' if p < 0.05 else '')))\n",
    "    )\n",
    "    \n",
    "    # Reshape the results to a more readable format\n",
    "    pivot_df = results_df.pivot(\n",
    "        index='cell_type',\n",
    "        columns='feature',\n",
    "        values=['direction','p_value', 'p_value_corrected', 'significance']\n",
    "    )\n",
    "    \n",
    "    return results_df, pivot_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4485506d-3139-45b8-b4ab-ddb0a8d19184",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['PRIME_score']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'anti_maa_hi'\n",
    "\n",
    "\n",
    "df = sc.get.obs_df(adata_amaa, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "# #Add a dummy expression values 0 if there are not cells in a given category\n",
    "# for cell_type in cell_list:\n",
    "\n",
    "#     if len(df[(df['majority_voting']==cell_type)&(df['anti_maa_hi']==True)])<2:\n",
    "   \n",
    "#         d ={features[0]:[0,0.000001],'majority_voting':[cell_type,cell_type],'anti_maa_hi':[True,True]}\n",
    "#         df = pd.concat((df,pd.DataFrame(data=d)))\n",
    "\n",
    "#     if len(df[(df['majority_voting']==cell_type)&(df['anti_maa_hi']==False)])<2:\n",
    "\n",
    "#         d ={features[0]:[0,0.000001],'majority_voting':[cell_type,cell_type],'anti_maa_hi':[False,False]}\n",
    "#         df = pd.concat((df,pd.DataFrame(data=d)))\n",
    "\n",
    "\n",
    "plt.figure(figsize=(12,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='box', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='Anti-MAA',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.25 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='right' )\n",
    "plt.ylim([0,2.5])\n",
    "plt.xlim([-1.0,28])\n",
    "\n",
    "handles = []\n",
    "\n",
    "## FIX THE COLORS TO CORRESPOND UMAPS \n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['op_tab24']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    \n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('PRIME_score_aMAA_all_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8f67794-aadb-4738-a046-8de876affa49",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.tsne(adata_amaa[adata_amaa.obs['anti_maa_hi']==False], color = ['PRIME_score'], cmap = 'coolwarm' , vmin=0,vmax=1., save='_PRIME_score_all_cells_aMAA_low.pdf')\n",
    "sc.pl.tsne(adata_amaa[adata_amaa.obs['anti_maa_hi']==True], color = ['PRIME_score'], cmap = 'coolwarm' , vmin=0,vmax=1., save='_PRIME_score_all_cells_aMAA_hi.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ceefc04-2852-4c8a-9d25-c6858432bfe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#SLAMF7 alone\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['SLAMF7']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'anti_maa_hi'\n",
    "\n",
    "df = sc.get.obs_df(adata_amaa, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "# #Add a dummy expression values 0 if there are not cells in a given category\n",
    "# for cell_type in cell_list:\n",
    "\n",
    "#     if len(df[(df['majority_voting']==cell_type)&(df['anti_maa_hi']==True)])<2:\n",
    "   \n",
    "#         d ={features[0]:[0,0.000001],'majority_voting':[cell_type,cell_type],'anti_maa_hi':[True,True]}\n",
    "#         df = pd.concat((df,pd.DataFrame(data=d)))\n",
    "\n",
    "#     if len(df[(df['majority_voting']==cell_type)&(df['anti_maa_hi']==False)])<2:\n",
    "\n",
    "#         d ={features[0]:[0,0.000001],'majority_voting':[cell_type,cell_type],'anti_maa_hi':[False,False]}\n",
    "#         df = pd.concat((df,pd.DataFrame(data=d)))\n",
    "\n",
    "\n",
    "plt.figure(figsize=(12,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='box', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='Anti-MAA',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.25 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  asterisk, ha='center', va='bottom', fontsize=12)\n",
    "        ax.text(idx, y_position-0.15, direction, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0,4.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='right' )\n",
    "plt.ylim([-0,5.5])\n",
    "plt.xlim([-1.0,28])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('SLAMF7_aMAA_all_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a2e42ef-888e-4357-8a3c-ab430afc1673",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.tsne(adata_amaa[adata_amaa.obs['anti_maa_hi']==False], color = ['SLAMF7'], cmap = 'coolwarm' , vmin=0,vmax=3.5, save='_SLAMF7_all_cells_aMAA_low.pdf')\n",
    "sc.pl.tsne(adata_amaa[adata_amaa.obs['anti_maa_hi']==True], color = ['SLAMF7'], cmap = 'coolwarm' , vmin=0,vmax=3.5, save='_SLAMF7_all_cells_aMAA_hi.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e77810d-b05e-44dd-adde-6970ad3a3f6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#TYMP\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['TYMP']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'anti_maa_hi'\n",
    "\n",
    "\n",
    "df = sc.get.obs_df(adata_amaa, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "# #Add a dummy expression values 0 if there are not cells in a given category\n",
    "# for cell_type in cell_list:\n",
    "\n",
    "#     if len(df[(df['majority_voting']==cell_type)&(df['anti_maa_hi']==True)])<2:\n",
    "   \n",
    "#         d ={features[0]:[0,0.000001],'majority_voting':[cell_type,cell_type],'anti_maa_hi':[True,True]}\n",
    "#         df = pd.concat((df,pd.DataFrame(data=d)))\n",
    "\n",
    "#     if len(df[(df['majority_voting']==cell_type)&(df['anti_maa_hi']==False)])<2:\n",
    "\n",
    "#         d ={features[0]:[0,0.000001],'majority_voting':[cell_type,cell_type],'anti_maa_hi':[False,False]}\n",
    "#         df = pd.concat((df,pd.DataFrame(data=d)))\n",
    "\n",
    "\n",
    "plt.figure(figsize=(12,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='box', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='Anti-MAA',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.25 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0,4.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='right' )\n",
    "plt.ylim([0,5.5])\n",
    "plt.xlim([-1.0,28])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('TYMP_aMAA_all_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78bf7981-75df-4d0c-8419-7f902c284422",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.tsne(adata_amaa[adata_amaa.obs['anti_maa_hi']==False], color = ['TYMP'], cmap = 'coolwarm' , vmin=0,vmax=3.5, save='TYMP_all_cells_aMAA_low.pdf')\n",
    "sc.pl.tsne(adata_amaa[adata_amaa.obs['anti_maa_hi']==True], color = ['TYMP'], cmap = 'coolwarm' , vmin=0,vmax=3.5, save='TYMP_all_cells_aMAA_hi.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9324e353-21ed-403d-8352-9593a5c2bebe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CD74\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['CD74']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'anti_maa_hi'\n",
    "\n",
    "df = sc.get.obs_df(adata_amaa, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "# #Add a dummy expression values 0 if there are not cells in a given category\n",
    "# for cell_type in cell_list:\n",
    "\n",
    "#     if len(df[(df['majority_voting']==cell_type)&(df['anti_maa_hi']==True)])<2:\n",
    "   \n",
    "#         d ={features[0]:[0.,0.000001,0.,0.000001],\n",
    "#             'majority_voting':[cell_type,cell_type,cell_type,cell_type],'anti_maa_hi':[False,False,True,True]}\n",
    "#         df = pd.concat((df,pd.DataFrame(data=d)))\n",
    "\n",
    "#     if len(df[(df['majority_voting']==cell_type)&(df['anti_maa_hi']==False)])<2:\n",
    "\n",
    "#         d ={features[0]:[0.,0.000001,0.,0.000001],\n",
    "#             'majority_voting':[cell_type,cell_type,cell_type,cell_type],'anti_maa_hi':[False,False,True,True]}\n",
    "#         df = pd.concat((df,pd.DataFrame(data=d)))\n",
    "\n",
    "\n",
    "plt.figure(figsize=(12,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='box', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='Anti-MAA',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.25 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0,4.0,5.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='right' )\n",
    "plt.ylim([0,6.])\n",
    "plt.xlim([-1.0,28])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('CD74_aMAA_all_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "708bf8bb-3637-4e17-8b02-b76f95688f46",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.tsne(adata_amaa[adata_amaa.obs['anti_maa_hi']==False], color = ['CD74'], cmap = 'coolwarm' , vmin=0,vmax=5.5, save='CD74_all_cells_aMAA_low.pdf')\n",
    "sc.pl.tsne(adata_amaa[adata_amaa.obs['anti_maa_hi']==True], color = ['CD74'], cmap = 'coolwarm' , vmin=0,vmax=5.5, save='CD74_all_cells_aMAA_hi.pdf')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c7862cf-f224-49a5-b0be-f52400cb935c",
   "metadata": {},
   "source": [
    "#### Immune regulators"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55bb0b5e-2bb3-42e2-89c8-d8bfc09e7438",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_list = ['PDCD1','LAG3','TMEM173','SIGLEC10','LILRB4','LILRB2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d098fcb-3d10-4455-910b-479021ad9086",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "for feature in reg_list:\n",
    "\n",
    "    features = [feature]\n",
    "    cell_key = 'majority_voting'\n",
    "    condition_key = 'anti_maa_hi'\n",
    "    \n",
    "    \n",
    "    df = sc.get.obs_df(adata_amaa, features+[cell_key,condition_key])\n",
    "    \n",
    "    df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "    df_all_cells = df.copy()\n",
    "    df_all_cells[cell_key] = 'Combined Cells'\n",
    "    \n",
    "    # Concatenate original and combined data\n",
    "    df = pd.concat([df, df_all_cells])\n",
    "    \n",
    "    cell_list=list(df[cell_key].unique())\n",
    "\n",
    "    # #Add a dummy expression values 0 if there are not cells in a given category\n",
    "    # for cell_type in cell_list:\n",
    "    \n",
    "    #     if ((len(df[(df['majority_voting']==cell_type)&(df['anti_maa_hi']==True)])<2)|\n",
    "    #         (len(df[(df['majority_voting']==cell_type)&(df['anti_maa_hi']==False)])<2)):\n",
    "       \n",
    "    #         d ={features[0]:[0.,0.0,0.,0.0],\n",
    "    #             'majority_voting':[cell_type,cell_type,cell_type,cell_type],'anti_maa_hi':[False,False,True,True]}\n",
    "    #         df = pd.concat((df,pd.DataFrame(data=d)))\n",
    "    \n",
    "    #     # if len(df[(df['majority_voting']==cell_type)&(df['anti_maa_hi']==False)])<2:\n",
    "    \n",
    "    #     #     d ={features[0]:[0.,0.000001,0.,0.000001],\n",
    "    #     #         'majority_voting':[cell_type,cell_type,cell_type,cell_type],'anti_maa_hi':[False,False,True,True]}\n",
    "    #     #     df = pd.concat((df,pd.DataFrame(data=d)))\n",
    "\n",
    "    \n",
    "    plt.figure(figsize=(12,3))\n",
    "    \n",
    "    ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='box', gap=.2, \n",
    "                              density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                            legend=True, linewidth=1,cut=0,\n",
    "                              alpha=0.5)\n",
    "    \n",
    "    ax.set(xlabel = None)\n",
    "    ax.legend(handles=ax.legend_.legend_handles, title='Anti-MAA',labels=['Low', 'High'])\n",
    "    \n",
    "    # Get the maximum y value for positioning asterisks\n",
    "    y_max = df.groupby(cell_key)[features[0]].max()\n",
    "    \n",
    "    # Add asterisks above each violin plot\n",
    "    results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "    \n",
    "    for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "\n",
    "        # Uncomment to add text only if there is a significance marker\n",
    "        # if asterisk:  \n",
    "        \n",
    "        y_position = y_max[cell_type] + 0.2 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "    plt.yticks([0.0,1.0,2.0,3.0,4.0,5.0])\n",
    "    plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='right' )\n",
    "    plt.ylim([0,6])\n",
    "    plt.xlim([-1.0,28])\n",
    "    \n",
    "    handles = []\n",
    "    \n",
    "    n_col =len(cell_list)\n",
    "    cmap = matplotlib.colormaps['tab10']\n",
    "    \n",
    "    # Take colors at regular intervals spanning the colormap.\n",
    "    colors = cmap(np.linspace(0, 1, n_col))\n",
    "    # colors='tab10'\n",
    "    \n",
    "    for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    # for ind in np.arange(len(cell_list)):\n",
    "\n",
    "        rgb = to_rgb(colors[4-ind // 2])\n",
    "       \n",
    "        if ind % 2 != 0:\n",
    "            rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "        violin.set_facecolor(rgb)\n",
    "        handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "    \n",
    "    # Adjust the top margin to make room for asterisks\n",
    "    plt.tight_layout()\n",
    "    plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "    \n",
    "    # plt.savefig(str(feature)+'_aMAA_all_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "    \n",
    "    plt.show()\n",
    "    \n",
    "    # results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49d30618-9223-4c76-ad1e-cc6fb8277c12",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CTL score\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['ctl_score']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'anti_maa_hi'\n",
    "\n",
    "df = sc.get.obs_df(adata_amaa, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "# #Add a dummy expression values 0 if there are not cells in a given category\n",
    "# for cell_type in cell_list:\n",
    "\n",
    "#     if len(df[(df['majority_voting']==cell_type)&(df['anti_maa_hi']==True)])<2:\n",
    "   \n",
    "#         d ={features[0]:[0.,0.000001,0.,0.000001],\n",
    "#             'majority_voting':[cell_type,cell_type,cell_type,cell_type],'anti_maa_hi':[False,False,True,True]}\n",
    "#         df = pd.concat((df,pd.DataFrame(data=d)))\n",
    "\n",
    "#     if len(df[(df['majority_voting']==cell_type)&(df['anti_maa_hi']==False)])<2:\n",
    "\n",
    "#         d ={features[0]:[0.,0.000001,0.,0.000001],\n",
    "#             'majority_voting':[cell_type,cell_type,cell_type,cell_type],'anti_maa_hi':[False,False,True,True]}\n",
    "#         df = pd.concat((df,pd.DataFrame(data=d)))\n",
    "\n",
    "\n",
    "plt.figure(figsize=(12,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='box', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='Anti-MAA',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.25 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0,4.0,5.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='right' )\n",
    "plt.ylim([-2,6.])\n",
    "plt.xlim([-1.0,28])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('CTL_aMAA_all_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38e38c16-a62a-4247-9ec3-07c95d1580b2",
   "metadata": {},
   "source": [
    "## CTL split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "537b282e-d44e-446d-a4ac-277944417652",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_th_naive.obs['ctl_hi'] = adata_th_naive.obs['ctl_score_pt']>np.quantile(np.unique(adata_th_naive.obs['ctl_score_pt']), [0.25,0.5,0.75])[1]\n",
    "adata_th_naive.obs['ctl_q1'] = adata_th_naive.obs['ctl_score_pt']>np.quantile(np.unique(adata_th_naive.obs['ctl_score_pt']), [0.25,0.5,0.75])[2]\n",
    "adata_th_naive.obs['ctl_q4'] = adata_th_naive.obs['ctl_score_pt']<=np.quantile(np.unique(adata_th_naive.obs['ctl_score_pt']), [0.25,0.5,0.75])[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6476785f-ed5a-4eb4-9d7c-0b905f4824d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PRIME\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['PRIME_score']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'ctl_hi'\n",
    "\n",
    "\n",
    "df = sc.get.obs_df(adata_th_naive, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "plt.figure(figsize=(12,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='box', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='CTL',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.25 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='right' )\n",
    "plt.ylim([-0.25,2.5])\n",
    "plt.xlim([-1.0,28])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('PRIME_CTL_all_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c9cfe02-0da3-49df-ba0f-75a92a5d4455",
   "metadata": {},
   "outputs": [],
   "source": [
    "#SLAMF7\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['SLAMF7']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'ctl_hi'\n",
    "\n",
    "\n",
    "df = sc.get.obs_df(adata_th_naive, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "plt.figure(figsize=(12,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='box', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='CTL',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.2 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='right' )\n",
    "plt.ylim([-0.25,5.5])\n",
    "plt.xlim([-1.0,28])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('SLAMF7_CTL_all_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5670c64-aa3a-4406-93a4-3dfc791817e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#TYMP\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['TYMP']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'ctl_hi'\n",
    "\n",
    "\n",
    "df = sc.get.obs_df(adata_th_naive, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "plt.figure(figsize=(12,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='box', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='CTL',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.2 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0,4.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='right' )\n",
    "plt.ylim([-0.25,5.5])\n",
    "plt.xlim([-1.0,28])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('TYMP_CTL_all_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58c331eb-63e2-4bad-ac84-d393f1dcb505",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CD74\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['CD74']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'ctl_hi'\n",
    "\n",
    "\n",
    "df = sc.get.obs_df(adata_th_naive, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "plt.figure(figsize=(12,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='box', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "# ax.legend(handles=ax.legend_.legendHandles, title='CTL',labels=['Low', 'High'])\n",
    "handles = ax.get_legend_handles_labels()[0]  # gets all handles\n",
    "ax.legend(handles=handles, labels=['Low', 'High'], title='CTL')\n",
    "\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.25 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='right' )\n",
    "plt.ylim([-0.25,5.5])\n",
    "plt.xlim([-1.0,28])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('CD74_CTL_all_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "602331b3-d384-430b-9c3b-d31468590289",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Double check: Cell CTL score split by median patient CTL score\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['ctl_score']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'ctl_hi'\n",
    "\n",
    "\n",
    "df = sc.get.obs_df(adata_th_naive, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "plt.figure(figsize=(12,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='box', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "# ax.legend(handles=ax.legend_.legendHandles, title='CTL',labels=['Low', 'High'])\n",
    "handles = ax.get_legend_handles_labels()[0]  # gets all handles\n",
    "ax.legend(handles=handles, labels=['Low', 'High'], title='CTL')\n",
    "\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.2 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='right' )\n",
    "plt.ylim([-2.25,5.5])\n",
    "plt.xlim([-1.0,28])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('cellCTL_by_ptCTL_all_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4aef6a2-094c-4e53-8845-b52987fda17c",
   "metadata": {},
   "source": [
    "## DC, MPhages and monocytes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e429f69-0122-4527-abfa-3529934c78dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_subset = adata_amaa[adata_amaa.obs['leiden'].isin(['r0','r1','r2','r3','r4','r5','r6','r7','r8','16','17'])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d0e8d91-f230-411e-9d71-660b086cce58",
   "metadata": {},
   "outputs": [],
   "source": [
    "dc_sig_genes = [\"SLC7A7\", \"LILRB2\", \"TNFAIP2\", \"SECTM1\", \"CSF2RB\", \"LST1\", \"BLNK\", \"LAT2\", \"FGR\", \"CTSS\", \"TYMP\", \"AOAH\", \"CD4\", \"PIK3AP1\", \"POU2F2\", \"HLA-DPA1\", \"ITGB2\", \"MZB1\", \"CD84\", \"HCLS1\", \"TNFRSF1B\", \"HAVCR2\", \"DOK2\", \"CD53\", \"SELL\", \"SLAMF7\", \"LCP1\", \"TAPBP\", \"RCSD1\", \"ARHGAP30\", \"CD37\", \"PSMB10\", \"EPSTI1\", \"CCDC69\"]\n",
    "sc.tl.score_genes(adata_subset, dc_sig_genes, score_name='dc_sig_score')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d13be4a-9b20-453f-abba-9ffd13890ed6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell type predictions by Celltypist\n",
    "predictions_subset = celltypist.annotate(adata_subset, model = 'Immune_All_High.pkl', majority_voting = True)\n",
    "print(predictions_subset.predicted_labels)\n",
    "adata_subset = predictions_subset.to_adata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60457641-7fcd-437d-8e65-53133d84f2a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "features = ['PRIME_score']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'anti_maa_hi'\n",
    "\n",
    "df = sc.get.obs_df(adata_subset, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "# n_col =len(cell_list)\n",
    "# cmap = matplotlib.colormaps['tab10']\n",
    "# colors = cmap(np.linspace(0., 1, 4))\n",
    "\n",
    "# Alphabetical order\n",
    "colors = [(0.5803921568627451, 0.403921568627451, 0.7411764705882353),\n",
    "          (0.7372549019607844, 0.7411764705882353, 0.1333333333),\n",
    "          (0.8901960784313725, 0.4666666666666667, 0.7607843137254902),\n",
    "          (0.17254901960784313, 0.6274509803921569, 0.17254901960784313)]\n",
    "\n",
    "subset_cmap = ListedColormap(colors, name='cmap1' )\n",
    "matplotlib.colormaps.register(name='cmap1', cmap=subset_cmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "073bd680-af45-4ae4-84f1-a4587a960b21",
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_cmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab90f35f-b808-45cb-b35d-f7200d7086b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## All DC, Mphages and monocytes\n",
    "sc.pl.tsne(adata_subset, color = 'majority_voting', palette ='cmap1',legend_loc = 'on data', alpha=0.5 , \n",
    "           add_outline=True, outline_width=(0.1,0.2), save='Li_DC_subset_CT_Immune_All_High.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62b934b3-bad5-44da-881d-8290bf57477f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Anti-MAA high\n",
    "print(adata_amaa[adata_amaa.obs['anti_maa_hi']==True].obs['patient'].unique())\n",
    "sc.pl.tsne(adata_subset[adata_subset.obs['anti_maa_hi']==True], color = 'majority_voting', palette ='cmap1',legend_loc = 'on data', alpha=0.5 , \n",
    "           add_outline=True, outline_width=(0.1,0.2), save='Li_DC_subset_CT_Immune_All_High_aMAAhi.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a984b7a4-dcdf-405d-a891-3bdd4b8a539f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Anti-MAA low\n",
    "\n",
    "print(adata_amaa[adata_amaa.obs['anti_maa_hi']==False].obs['patient'].unique())\n",
    "\n",
    "sc.pl.tsne(adata_subset[adata_subset.obs['anti_maa_hi']==False], color = 'majority_voting', palette ='cmap1',legend_loc = 'on data', alpha=0.5 , \n",
    "           add_outline=True, outline_width=(0.1,0.2), save='Li_DC_subset_CT_Immune_All_High_aMAAlow.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b174a28a-76b7-436c-9dc0-868c211388bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['PRIME_score']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'anti_maa_hi'\n",
    "\n",
    "df = sc.get.obs_df(adata_subset, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "plt.figure(figsize=(6,2))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='point', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.set(ylabel = 'PRIME score')\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='Anti-MAA',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.4 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  asterisk, ha='center', va='bottom', fontsize=12)\n",
    "        ax.text(idx, y_position-0.1, direction, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=15,  fontsize=12, ha='center' )\n",
    "plt.ylim([0,2.5])\n",
    "plt.xlim([-0.5,5.3])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('PRIME_score_aMAA_Select_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ef7b570-94e1-49a9-b16e-7d9257e76f55",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['SLAMF7']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'anti_maa_hi'\n",
    "\n",
    "df = sc.get.obs_df(adata_subset, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "plt.figure(figsize=(6,2))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='point', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.set(ylabel = 'SLAMF7')\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='Anti-MAA',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 1.2 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  asterisk, ha='center', va='bottom', fontsize=12)\n",
    "        ax.text(idx, y_position-0.15, direction, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0,4.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=15,  fontsize=12, ha='center' )\n",
    "plt.ylim([0,5])\n",
    "plt.xlim([-0.5,5.3])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('SLAMF7_aMAA_Select_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "312ea531-6098-41f3-ba65-476497390684",
   "metadata": {},
   "outputs": [],
   "source": [
    "### sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['TYMP']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'anti_maa_hi'\n",
    "\n",
    "df = sc.get.obs_df(adata_subset, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "plt.figure(figsize=(6,2))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='point', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.set(ylabel = 'TYMP')\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='Anti-MAA',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.5 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  asterisk, ha='center', va='bottom', fontsize=12)\n",
    "        ax.text(idx, y_position-0.15, direction, ha='center', va='bottom', fontsize=12)\n",
    "        \n",
    "plt.yticks([0.0,1.0,2.0,3.0,4.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=15,  fontsize=12, ha='center' )\n",
    "plt.ylim([0,5])\n",
    "plt.xlim([-0.5,5.3])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('TYMP_aMAA_Select_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a000169b-8568-4274-a482-6399aa0cc99d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['CD74']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'anti_maa_hi'\n",
    "\n",
    "df = sc.get.obs_df(adata_subset, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "plt.figure(figsize=(6,2))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='point', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.set(ylabel = 'CD74')\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='Anti-MAA',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 1.5 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  asterisk, ha='center', va='bottom', fontsize=12)\n",
    "        ax.text(idx, y_position-0.15, direction, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0,4.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=15,  fontsize=13, ha='center' )\n",
    "plt.ylim([0,5.5])\n",
    "plt.xlim([-0.5,5.3])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('CD74_aMAA_Select_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4eb37cc-6057-4c84-af66-d88c04328d7b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "68e5a431-5a3b-4d98-901e-f00b36c3eeea",
   "metadata": {},
   "source": [
    "### CTL split, select cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7ff32de-11ea-4b92-80e3-561f86496b4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_subset = adata_th_naive[adata_th_naive.obs['leiden'].isin(['r0','r1','r2','r3','r4','r5','r6','r7','r8','16','17'])].copy()\n",
    "# Cell type predictions by Celltypist\n",
    "predictions_subset = celltypist.annotate(adata_subset, model = 'Immune_All_High.pkl', majority_voting = True)\n",
    "print(predictions_subset.predicted_labels)\n",
    "adata_subset = predictions_subset.to_adata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1da94d7-ab8a-4928-b519-0e7b1081f6f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['PRIME_score']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'ctl_hi'\n",
    "\n",
    "df = sc.get.obs_df(adata_subset, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "plt.figure(figsize=(9,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='point', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='CTL',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.25 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='center' )\n",
    "plt.ylim([0,2.5])\n",
    "plt.xlim([-0.5,5])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('PRIME_score_CTL_Select_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3d9b9ad-bdf4-46cf-973e-ccb75ede2794",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['SLAMF7']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'ctl_hi'\n",
    "\n",
    "df = sc.get.obs_df(adata_subset, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "plt.figure(figsize=(9,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='point', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='CTL',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.25 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0,4.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='center' )\n",
    "plt.ylim([0,5])\n",
    "plt.xlim([-0.5,5])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('SLAMF7_CTL_Select_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4706dfa6-f005-49a7-87c2-8af5c16f468a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['TYMP']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'ctl_hi'\n",
    "\n",
    "df = sc.get.obs_df(adata_subset, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "plt.figure(figsize=(9,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='point', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='CTL',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.25 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='center' )\n",
    "plt.ylim([0,5])\n",
    "plt.xlim([-0.5,5])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('TYMP_CTL_Select_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "428ba432-6895-413c-ad46-f5edcd75f641",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(font_scale=1.5)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "features = ['CD74']\n",
    "cell_key = 'majority_voting'\n",
    "condition_key = 'ctl_hi'\n",
    "\n",
    "df = sc.get.obs_df(adata_subset, features+[cell_key,condition_key])\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "plt.figure(figsize=(9,3))\n",
    "\n",
    "ax=sns.violinplot(data=df, x=\"majority_voting\", y=features[0], hue=condition_key, split=True, inner='point', gap=.2, \n",
    "                          density_norm='width', width=0.8, palette=['.4', '.7'], order=cell_list,\n",
    "                        legend=True, linewidth=1,cut=0,\n",
    "                          alpha=0.5)\n",
    "\n",
    "ax.set(xlabel = None)\n",
    "ax.legend(handles=ax.legend_.legendHandles, title='CTL',labels=['Low', 'High'])\n",
    "\n",
    "# Get the maximum y value for positioning asterisks\n",
    "y_max = df.groupby(cell_key)[features[0]].max()\n",
    "\n",
    "# Add asterisks above each violin plot\n",
    "results_long, results_pivot = compare_gene_expression(df, cell_list, features, condition_key, False, True, cell_key )\n",
    "\n",
    "for idx, (cell_type, direction, asterisk) in enumerate(zip(results_long['cell_type'], results_long['direction'], results_long['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[cell_type] + 0.25 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.yticks([0.0,1.0,2.0,3.0])\n",
    "plt.xticks(np.arange(len(cell_list)),cell_list,rotation=30,  fontsize=10, ha='center' )\n",
    "plt.ylim([0,5])\n",
    "plt.xlim([-0.5,5])\n",
    "\n",
    "handles = []\n",
    "\n",
    "n_col =len(cell_list)\n",
    "cmap = matplotlib.colormaps['tab10']\n",
    "\n",
    "# Take colors at regular intervals spanning the colormap.\n",
    "colors = cmap(np.linspace(0, 1, n_col+1))\n",
    "# colors='tab10'\n",
    "\n",
    "for ind, violin in enumerate(ax.findobj(PolyCollection)):\n",
    "    rgb = to_rgb(colors[4-ind // 2])\n",
    "   \n",
    "    if ind % 2 != 0:\n",
    "        rgb = 0.5 + 0.5 * np.array(rgb)  # make whiter\n",
    "    violin.set_facecolor(rgb)\n",
    "    handles.append(plt.Rectangle((0, 0), 0, 0, facecolor=rgb, edgecolor='black'))\n",
    "\n",
    "# Adjust the top margin to make room for asterisks\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(top=1.5)  # Adjust this value if needed\n",
    "\n",
    "plt.savefig('CD74_CTL_Select_cells_Li.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "results_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21ae566d-5723-4792-997c-81e14f8d7431",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
